sql/2020/03_Markup/pages_wpt_bodies_by_device_and_protocol.sql

```diff
@@ -0,0 +1,46 @@
+#standardSQL
+# page wpt_bodies metrics grouped by device
+
+# helper to create percent fields
+CREATE TEMP FUNCTION AS_PERCENT (freq FLOAT64, total FLOAT64) RETURNS FLOAT64 AS (
+  ROUND(SAFE_DIVIDE(freq, total), 4)
+);
+
+# returns all the data we need from _wpt_bodies
+CREATE TEMPORARY FUNCTION get_wpt_bodies_protocols(wpt_bodies_string STRING)
+RETURNS ARRAY<STRING> LANGUAGE js AS '''
+var result = [];
+try {
+    var wpt_bodies = JSON.parse(wpt_bodies_string); 
+
+    if (Array.isArray(wpt_bodies) || typeof wpt_bodies != 'object') return result;
+
+    if (wpt_bodies.anchors && wpt_bodies.anchors.rendered && wpt_bodies.anchors.rendered.protocols) {
+        return Object.keys(wpt_bodies.anchors.rendered.protocols);
+    }
+
+} catch (e) {}
+return result;
+''';
+
+SELECT
+  _TABLE_SUFFIX AS client,
+  COUNT(DISTINCT url) AS pages,
+  total,
+  protocol,
```

Comment:
Casefold. Also LIMIT/HAVING if this is high cardinality. Probably not.
```suggestion
  LOWER(protocol) AS protocol,
```

Reply:
Already lowercase in the custom metric. Added a LIMIT.